hash_practice/exercises.py

```diff
@@ -2,18 +2,61 @@
 def grouped_anagrams(strings):
     """ This method will return an array of arrays.
         Each subarray will have strings which are anagrams of each other
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n) -> where n equals the total number of characters in all the words
+        Space Complexity: O(n)
     """
-    pass
+    
+    anagrams = {}
+    for word in strings:
+        alphabet = [0] * 26
+        for letter in word:
+            alphabet[ord(letter.lower()) - ord("a")] += 1
+        if tuple(alphabet) in anagrams:
+            anagrams[tuple(alphabet)].append(word)
+        else: 
+            anagrams[tuple(alphabet)] = [word]
+    return_list = []
+    for k, v in anagrams.items():
+        return_list.append(v)
+    return return_list
+
+
 
 def top_k_frequent_elements(nums, k):
     """ This method will return the k most common elements
         In the case of a tie it will select the first occuring element.
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n) --> Reduced from O(3n)
+        Space Complexity: O(n)
     """
-    pass
+    if len(nums) <= 1:
+        return nums
+
+    freq_table = {}
+    for num in nums:
+        if num not in freq_table:
+            freq_table[num] = 0
+        freq_table[num] += 1
+
+    return_list = [None] * len(nums)

```

Comment:
In the same manner this list is going to be n items long, but most of the entries will likely be 0, so it's also very sparse and can likely be optimized a little. (It will still be O(n).)

For example, the instructor solution I'm looking at uses the [heapq library](https://docs.python.org/3/library/heapq.html) to maintain a "heap". In short, a heap keeps the smallest value in easy access, allowing it to be popped off, making this a good solution for this top-k type problem.

Now, I'm not saying you needed to go so far afield. I don't think I would have thought of the heap solution myself, honestly. I would probably just have converted the key-value pairs of `freq_table` into a list of tuples and then sorted on the value, then dropped down to k pairs. Still kind of ugly, but much less sparse, and likely less space used.

Anyway, I just wanted to point out a similar place where you were using sparse data. But don't sweat this too much, your solution is fine.